pkg/export/gcm/export_gcm_test.go

```diff
@@ -0,0 +1,100 @@
+package gcm_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/GoogleCloudPlatform/prometheus-engine/pkg/export/gcm/promtest"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func TestExport_CounterReset(t *testing.T) {
+	const interval = 30 * time.Second
+
+	prom := promtest.Prometheus("quay.io/prometheus/prometheus:v2.47.2")
+	export := promtest.LocalExportWithGCM(promtest.GCMServiceAccountOrFail(t))
+
+	it := promtest.NewIngestionTest(t, []promtest.Backend{prom, export})
+
+	counter := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "pe_test_counter_total",
+		Help: "Test counter used by prometheus-engine export GCM acceptance tests.",
+	}, []string{"foo"})
+	var c prometheus.Counter
+
+	it.RecordScrapes(func(r prometheus.Registerer, scrape promtest.Scrape) {
+		r.MustRegister(counter)
+
+		// No metric.
+		scrape(interval)
+
+		c = counter.WithLabelValues("bar")
+		c.Add(200)
+
+		scrape(interval).
+			Expect(200, c, prom)
+		// Nothing is expected for GMP due to cannibalization.
+		// TODO(bwplotka): Fix with b/259261536.
+
+		c.Add(10)
+		scrape(interval).
+			Expect(10, c, export).
+			Expect(210, c, prom)
+
+		c.Add(50)
+		scrape(interval).
+			Expect(60, c, export).
+			Expect(260, c, prom)
+
+		// Reset to 0, then add something that still should indicate "decreased counter"
+		// in absolute values, but our current cannibalization logic is wrong here.
+		// It's perhaps unusual that reset happens without scrape target restart
+		// so, it was easy to forget about this case.
+		counter.Reset()
```

Comment:
Great find on the bug!

It doesn't look like you have a case that just resets without adding. Would it make sense to add one?

Reply:
👍🏽 

Any feedback on test clarity?

Would it be easier to read if we have scenario of scrapes/metrics and expect certain samples PER scrape at the end (instead per scrape)? 🤗

Reply:
I like the granularity you have right now! I feel like it's more low-level so we can always create a utility that wraps this one and does checking at the end.